lib/de.ml

```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+            + (c.precode_freqs.(sym) * (extra + c.precode_lens.(sym)))
+        done
+      ; for sym = 0 to 255 do
+          dynamic_cost :=
+            !dynamic_cost + (c.freqs.litlen.(sym) * c.codes.lens.litlen.(sym))
+        done
+      ; for sym = 0 to 143 do
+          static_cost := !static_cost + (c.freqs.litlen.(sym) * 8)
+        done
+      ; for sym = 144 to 255 do
+          static_cost := !static_cost + (c.freqs.litlen.(sym) * 9)
+        done
+      ; dynamic_cost := !dynamic_cost + c.codes.lens.litlen.(256)
+      ; static_cost := !static_cost + 7
+      ; for sym = 257 to 257 + Array.length extra_length_bits - 1 do
+          let extra = extra_length_bits.(sym - 257) in
+          dynamic_cost :=
+            !dynamic_cost
+            + (c.freqs.litlen.(sym) * (extra + c.codes.lens.litlen.(sym)))
+          ; static_cost :=
+              !static_cost
+              + c.freqs.litlen.(sym)
+                * (extra + c.static_codes.lens.litlen.(sym))
+        done
+      ; for sym = 0 to Array.length extra_offset_bits - 1 do
+          let extra = extra_offset_bits.(sym) in
+          dynamic_cost :=
+            !dynamic_cost
+            + (c.freqs.offset.(sym) * (extra + c.codes.lens.offset.(sym)))
+          ; static_cost := !static_cost + (c.freqs.offset.(sym) * (extra + 5))
+        done
+      ; uncompressed_cost :=
+          !uncompressed_cost
+          + (-(os.bits + 3) land 7)
+          + 32
+          + (40 * (((block_length + 65535 - 1) / 65535) - 1))
+          + (8 * block_length)
+      ; let block_type, codes =
+          if !dynamic_cost < min !static_cost !uncompressed_cost then
+            blocktype_dynamic_huffman, Some c.codes
+          else if !static_cost < !uncompressed_cost then
+            blocktype_static_huffman, Some c.static_codes
+          else blocktype_uncompressed, None in
+        if block_type = blocktype_uncompressed then begin
+          os.i_pos <- !block_begin
+          ; write_uncompressed_blocks os block_length is_final_block
+        end
+        else begin
+          write_block_header os is_final_block block_type
+          ; if block_type = blocktype_dynamic_huffman then
+              write_huffman_header c os
+          ; write_sequences os (Option.get codes) sequences block block_begin
+          ; write_end_of_block os (Option.get codes)
+        end
+
+    let init_block_split_stats stats =
+      Array.fill stats.new_observations 0 num_observation_types 0
+      ; Array.fill stats.observations 0 num_observation_types 0
+      ; stats.num_new_observations <- 0
+      ; stats.num_observations <- 0
+
+    let reset_symbol_frequencies c =
+      Array.fill c.freqs.litlen 0 (Array.length c.freqs.litlen) 0
+      ; Array.fill c.freqs.offset 0 (Array.length c.freqs.offset) 0
+
+    let finish_sequence seq litrunlen = seq.litrunlen_and_length <- litrunlen
+    let num_observations_per_block_check = 512
+
+    let do_end_block_check stats block_length =
```

Comment:
What is the purpose of this function? It tries to check if the input has a block tagged with last?

Reply:
This function contains some heuristic for choosing if a new block should be created or not